File: python/src/warehouse/adapters/repository.py
"""In-memory order repository keyed by immutable order id."""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING

from warehouse.domain.errors import PersistenceConflict

if TYPE_CHECKING:
    from warehouse.domain.order import Order, OrderId


class InMemoryOrderRepository:
    """OrderRepository double keeping snapshots, never aliases."""

    def __init__(self) -> None:
        """Start with an empty store."""
        self._lock = threading.Lock()
        self._orders: dict[OrderId, Order] = {}
        self._by_key: dict[str, tuple[str, Order]] = {}
        self.saved: list[Order] = []
        self.fail_save = False

    def save(self, order: Order, expected_version: int) -> Order | PersistenceConflict:
        """Store a snapshot under compare-and-set version rules."""
        with self._lock:
            if self.fail_save:
                return PersistenceConflict(
                    f"forced save failure for {order.id.value}",
                )
            current = self._orders.get(order.id)
            current_version = current.version if current is not None else 0
            if current_version != expected_version:
                return PersistenceConflict(
                    f"version conflict for {order.id.value}: "
                    f"expected {expected_version}, stored {current_version}",
                )
            snapshot = order.snapshot()
            snapshot.bump_version()
            self._orders[order.id] = snapshot
            self.saved.append(snapshot.snapshot())
            return snapshot.snapshot()

    def remember_idempotency(self, key: str, fingerprint: str, order: Order) -> None:
        """Record a successful command so retries can replay the snapshot."""
        with self._lock:
            self._by_key[key] = (fingerprint, order.snapshot())

    def get_by_idempotency_key(self, key: str) -> tuple[str, Order] | None:
        """Return the fingerprint and snapshot for a previous command."""
        with self._lock:
            found = self._by_key.get(key)
            if found is None:
                return None
            fingerprint, order = found
            return fingerprint, order.snapshot()

    def get(self, order_id: OrderId) -> Order | None:
        """Return a stored snapshot or None; an unknown id never raises."""
        with self._lock:
            stored = self._orders.get(order_id)
            if stored is None:
                return None
            return stored.snapshot()

File: typescript/src/adapters/repository.ts
/**
 * In-memory order repository keyed by immutable order id.
 */

import type { IdempotencyRecord, OrderRepository } from "../application/ports";
import { PersistenceConflict } from "../domain/errors";
import type { Order, OrderId } from "../domain/order";

/**
 * `OrderRepository` double keeping snapshots, never aliases.
 *
 * Lookup of an unknown id returns undefined and never raises (CONTRACTS.md
 * §2 binding clarification).
 */
export class InMemoryOrderRepository implements OrderRepository {
  private readonly orders = new Map<OrderId, Order>();

  private readonly byKey = new Map<string, IdempotencyRecord>();

  /**
   * When true, `save` returns `PersistenceConflict` without storing.
   */
  failSave = false;

  /**
   * Every snapshot returned from save, in call order — for assertions.
   */
  readonly saved: Order[] = [];

  /**
   * Stores a snapshot under compare-and-set version rules.
   *
   * @param order - the order to persist.
   * @param expectedVersion - the version the caller believes is stored.
   * @returns a detached snapshot, or `PersistenceConflict` on a lost race.
   */
  save(order: Order, expectedVersion: number): Order | PersistenceConflict {
    if (this.failSave) {
      return new PersistenceConflict(`forced save failure for ${order.id}`);
    }
    const current = this.orders.get(order.id);
    const currentVersion = current === undefined ? 0 : current.version;
    if (currentVersion !== expectedVersion) {
      return new PersistenceConflict(
        `version conflict for ${order.id}: expected ${String(expectedVersion)}, stored ${String(currentVersion)}`,
      );
    }
    const snapshot = order.snapshot();
    snapshot.bumpVersion();
    this.orders.set(order.id, snapshot);
    this.saved.push(snapshot.snapshot());
    return snapshot.snapshot();
  }

  /**
   * Records a successful command so retries can replay the snapshot.
   *
   * @param key - the idempotency key of the command.
   * @param fingerprint - stable payload identity of the command.
   * @param order - the persisted paid snapshot to replay.
   */
  rememberIdempotency(key: string, fingerprint: string, order: Order): void {
    this.byKey.set(key, { fingerprint, order: order.snapshot() });
  }

  /**
   * Returns the fingerprint and snapshot for a previous command.
   *
   * @param key - the idempotency key to look up.
   * @returns the record, or undefined when this key has not succeeded.
   */
  getByIdempotencyKey(key: string): IdempotencyRecord | undefined {
    const found = this.byKey.get(key);
    if (found === undefined) {
      return undefined;
    }
    return { fingerprint: found.fingerprint, order: found.order.snapshot() };
  }

  /**
   * Looks up an order by identifier; absence is undefined, never a throw.
   *
   * @param orderId - the identifier to look up.
   * @returns a detached snapshot, or undefined when absent.
   */
  get(orderId: OrderId): Order | undefined {
    const stored = this.orders.get(orderId);
    if (stored === undefined) {
      return undefined;
    }
    return stored.snapshot();
  }
}
